dbms/src/TiDB/Schema/SchemaSyncService.cpp

```diff
@@ -40,35 +40,42 @@ SchemaSyncService::SchemaSyncService(DB::Context & context_)
     handle = background_pool.addTask(
         [&, this] {
             String stage;
+            auto keyspaces = context.getTMTContext().getStorages().getAllKeyspaces();
+            KeyspaceID cur_ks = NullspaceID;
             bool done_anything = false;
             try
             {
-                /// Do sync schema first, then gc.
-                /// They must be performed synchronously,
-                /// otherwise table may get mis-GC-ed if RECOVER was not properly synced caused by schema sync pause but GC runs too aggressively.
-                // GC safe point must be obtained ahead of syncing schema.
                 auto gc_safe_point = PDClientHelper::getGCSafePointWithRetry(context.getTMTContext().getPDClient());
-                stage = "Sync schemas";
-                done_anything = syncSchemas();
-                if (done_anything)
-                    GET_METRIC(tiflash_schema_trigger_count, type_timer).Increment();
-
-                stage = "GC";
-                done_anything = gc(gc_safe_point);
-
+                for (auto const iter : keyspaces)
+                {
+                    auto ks = iter.first;
+                    LOG_DEBUG(log, "Auto sync schema for keyspace {}", ks);
```

Comment:
Is this expected that the schema sync task running in multiple thread？I got the following debug log with keyspace 24. Dose the task running too frequently if we have thousands of keyspace?

```
[2023/02/16 17:07:59.840 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=58]
[2023/02/16 17:08:00.078 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=54]
[2023/02/16 17:08:00.172 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=43]
[2023/02/16 17:08:00.198 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=31]
[2023/02/16 17:08:00.212 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=56]
[2023/02/16 17:08:00.251 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=55]
[2023/02/16 17:08:10.293 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=31]
[2023/02/16 17:08:10.371 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=43]
[2023/02/16 17:08:10.436 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=50]
[2023/02/16 17:08:10.454 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=49]
[2023/02/16 17:08:10.538 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=59]
[2023/02/16 17:08:10.574 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=54]
[2023/02/16 17:08:10.590 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=57]
c[2023/02/16 17:08:10.635 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=51]
[2023/02/16 17:08:10.666 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=56]
[2023/02/16 17:08:10.714 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=53]
[2023/02/16 17:08:10.902 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=60]
[2023/02/16 17:08:10.967 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=61]
[2023/02/16 17:08:11.009 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=55]
[2023/02/16 17:08:11.051 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=52]
[2023/02/16 17:08:11.086 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=62]
[2023/02/16 17:08:11.151 +08:00] [DEBUG] [SchemaSyncService.cpp:52] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=58]
```

Reply:
Oh, I think the task running in the BackgroundProcessingPool with 0ms interval, not parallel. `interval_ms` should be specified when adding the task.

Reply:
What is the value of `done_anything`? If done_anything is true, then the task will be executed again immediately.

Reply:
@JaySon-Huang I add some debug logs, and the `done_anything` is `false`, but the task seems runs again immediately.

```cpp
           try
            {
                LOG_DEBUG(log, "SchemaSyncService: start to sync schema and gc for {} keyspaces", keyspaces.size());
                auto gc_safe_point = PDClientHelper::getGCSafePointWithRetry(context.getTMTContext().getPDClient());
                for (auto const iter : keyspaces)
                {
                    auto ks = iter.first;
                    LOG_DEBUG(log, "Auto sync schema for keyspace {}", ks);
                    cur_ks = ks;
                    /// Do sync schema first, then gc.
                    /// They must be performed synchronously,
                    /// otherwise table may get mis-GC-ed if RECOVER was not properly synced caused by schema sync pause but GC runs too aggressively.
                    // GC safe point must be obtained ahead of syncing schema.
                    stage = "Sync schemas";
                    done_anything = syncSchemas(ks);
                    LOG_DEBUG(log, "SchemaSyncService: sync schema for keyspace {} done_anything {}", ks, done_anything);
                    if (done_anything)
                        GET_METRIC(tiflash_schema_trigger_count, type_timer).Increment();

                    stage = "GC";
                    done_anything = gc(gc_safe_point, ks);
                    LOG_DEBUG(log, "SchemaSyncService: gc for keyspace {} done_anything {}", ks, done_anything);
                }
                return done_anything;
            }
```

the logs

```
[2023/02/16 22:57:03.802 +08:00] [DEBUG] [SchemaSyncService.cpp:48] ["SchemaSyncService: start to sync schema and gc for 1 keyspaces"] [source=SchemaSyncService] [thread_id=47]
[2023/02/16 22:57:03.805 +08:00] [DEBUG] [SchemaSyncService.cpp:53] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=47]
[2023/02/16 22:57:03.807 +08:00] [DEBUG] [SchemaSyncService.cpp:61] ["SchemaSyncService: sync schema for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=47]
[2023/02/16 22:57:03.807 +08:00] [DEBUG] [SchemaSyncService.cpp:67] ["SchemaSyncService: gc for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=47]
[2023/02/16 22:57:03.906 +08:00] [DEBUG] [SchemaSyncService.cpp:48] ["SchemaSyncService: start to sync schema and gc for 1 keyspaces"] [source=SchemaSyncService] [thread_id=51]
[2023/02/16 22:57:03.909 +08:00] [DEBUG] [SchemaSyncService.cpp:53] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=51]
[2023/02/16 22:57:03.912 +08:00] [DEBUG] [SchemaSyncService.cpp:61] ["SchemaSyncService: sync schema for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=51]
[2023/02/16 22:57:03.912 +08:00] [DEBUG] [SchemaSyncService.cpp:67] ["SchemaSyncService: gc for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=51]
[2023/02/16 22:57:04.050 +08:00] [DEBUG] [SchemaSyncService.cpp:48] ["SchemaSyncService: start to sync schema and gc for 1 keyspaces"] [source=SchemaSyncService] [thread_id=45]
[2023/02/16 22:57:04.053 +08:00] [DEBUG] [SchemaSyncService.cpp:53] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=45]
[2023/02/16 22:57:04.055 +08:00] [DEBUG] [SchemaSyncService.cpp:61] ["SchemaSyncService: sync schema for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=45]
[2023/02/16 22:57:04.055 +08:00] [DEBUG] [SchemaSyncService.cpp:67] ["SchemaSyncService: gc for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=45]
[2023/02/16 22:57:04.060 +08:00] [DEBUG] [SchemaSyncService.cpp:48] ["SchemaSyncService: start to sync schema and gc for 1 keyspaces"] [source=SchemaSyncService] [thread_id=43]
[2023/02/16 22:57:04.062 +08:00] [DEBUG] [SchemaSyncService.cpp:53] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=43]
[2023/02/16 22:57:04.064 +08:00] [DEBUG] [SchemaSyncService.cpp:61] ["SchemaSyncService: sync schema for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=43]
[2023/02/16 22:57:04.064 +08:00] [DEBUG] [SchemaSyncService.cpp:67] ["SchemaSyncService: gc for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=43]
[2023/02/16 22:57:04.117 +08:00] [DEBUG] [SchemaSyncService.cpp:48] ["SchemaSyncService: start to sync schema and gc for 1 keyspaces"] [source=SchemaSyncService] [thread_id=40]
[2023/02/16 22:57:04.120 +08:00] [DEBUG] [SchemaSyncService.cpp:53] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=40]
[2023/02/16 22:57:04.122 +08:00] [DEBUG] [SchemaSyncService.cpp:61] ["SchemaSyncService: sync schema for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=40]
[2023/02/16 22:57:04.122 +08:00] [DEBUG] [SchemaSyncService.cpp:67] ["SchemaSyncService: gc for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=40]
[2023/02/16 22:57:04.203 +08:00] [DEBUG] [SchemaSyncService.cpp:48] ["SchemaSyncService: start to sync schema and gc for 1 keyspaces"] [source=SchemaSyncService] [thread_id=49]
[2023/02/16 22:57:04.206 +08:00] [DEBUG] [SchemaSyncService.cpp:53] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=49]
[2023/02/16 22:57:04.208 +08:00] [DEBUG] [SchemaSyncService.cpp:61] ["SchemaSyncService: sync schema for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=49]
[2023/02/16 22:57:04.208 +08:00] [DEBUG] [SchemaSyncService.cpp:67] ["SchemaSyncService: gc for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=49]
[2023/02/16 22:57:04.230 +08:00] [DEBUG] [SchemaSyncService.cpp:48] ["SchemaSyncService: start to sync schema and gc for 1 keyspaces"] [source=SchemaSyncService] [thread_id=48]
[2023/02/16 22:57:04.232 +08:00] [DEBUG] [SchemaSyncService.cpp:53] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=48]
[2023/02/16 22:57:04.235 +08:00] [DEBUG] [SchemaSyncService.cpp:61] ["SchemaSyncService: sync schema for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=48]
[2023/02/16 22:57:04.235 +08:00] [DEBUG] [SchemaSyncService.cpp:67] ["SchemaSyncService: gc for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=48]
[2023/02/16 22:57:04.318 +08:00] [DEBUG] [SchemaSyncService.cpp:48] ["SchemaSyncService: start to sync schema and gc for 1 keyspaces"] [source=SchemaSyncService] [thread_id=53]
[2023/02/16 22:57:04.321 +08:00] [DEBUG] [SchemaSyncService.cpp:53] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=53]
[2023/02/16 22:57:04.323 +08:00] [DEBUG] [SchemaSyncService.cpp:61] ["SchemaSyncService: sync schema for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=53]
[2023/02/16 22:57:04.323 +08:00] [DEBUG] [SchemaSyncService.cpp:67] ["SchemaSyncService: gc for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=53]
[2023/02/16 22:57:04.528 +08:00] [DEBUG] [SchemaSyncService.cpp:48] ["SchemaSyncService: start to sync schema and gc for 1 keyspaces"] [source=SchemaSyncService] [thread_id=44]
[2023/02/16 22:57:04.530 +08:00] [DEBUG] [SchemaSyncService.cpp:53] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=44]
[2023/02/16 22:57:04.533 +08:00] [DEBUG] [SchemaSyncService.cpp:61] ["SchemaSyncService: sync schema for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=44]
[2023/02/16 22:57:04.533 +08:00] [DEBUG] [SchemaSyncService.cpp:67] ["SchemaSyncService: gc for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=44]
[2023/02/16 22:57:04.621 +08:00] [DEBUG] [SchemaSyncService.cpp:48] ["SchemaSyncService: start to sync schema and gc for 1 keyspaces"] [source=SchemaSyncService] [thread_id=46]
[2023/02/16 22:57:04.625 +08:00] [DEBUG] [SchemaSyncService.cpp:53] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=46]
[2023/02/16 22:57:04.628 +08:00] [DEBUG] [SchemaSyncService.cpp:61] ["SchemaSyncService: sync schema for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=46]
[2023/02/16 22:57:04.628 +08:00] [DEBUG] [SchemaSyncService.cpp:67] ["SchemaSyncService: gc for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=46]
[2023/02/16 22:57:04.706 +08:00] [DEBUG] [SchemaSyncService.cpp:48] ["SchemaSyncService: start to sync schema and gc for 1 keyspaces"] [source=SchemaSyncService] [thread_id=34]
[2023/02/16 22:57:04.709 +08:00] [DEBUG] [SchemaSyncService.cpp:53] ["Auto sync schema for keyspace 24"] [source=SchemaSyncService] [thread_id=34]
[2023/02/16 22:57:04.711 +08:00] [DEBUG] [SchemaSyncService.cpp:61] ["SchemaSyncService: sync schema for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=34]
[2023/02/16 22:57:04.711 +08:00] [DEBUG] [SchemaSyncService.cpp:67] ["SchemaSyncService: gc for keyspace 24 done_anything false"] [source=SchemaSyncService] [thread_id=34]
```

BTW: Is it more reasonable to add tasks for every keyspace schma sync ? Or it will cause lots of useless schema sync attemps when it has thousands of keyspaces.

Reply:
https://github.com/pingcap/tiflash/pull/6910

Reply:
This pr to the source branch fixed potential useless sync schema attempts https://github.com/iosmanthus/tiflash/pull/1

@iosmanthus @JaySon-Huang PTAL~

Reply:
Fix by the pull request above.